File: app/src/main/java/app/we/go/emojidraw/api/EmojiDetectionRequest.java
package app.we.go.emojidraw.api;


import com.google.gson.annotations.SerializedName;

import java.util.Collections;
import java.util.List;

@SuppressWarnings("WeakerAccess")
public class EmojiDetectionRequest {

    private static final String OPTIONS = "enable_pre_space";
    private static final String LANGUAGE = "emoji_rotated";

    String options;

    List<Request> requests;

    private static class Request {
        @SerializedName("writing_guide")
        WritingGuide writingGuide;

        int[][][] ink;

        String language;

        private static class WritingGuide {
            @SerializedName("writing_area_width")
            public final String width;
            @SerializedName("writing_area_height")
            public final String height;

            public WritingGuide(String width, String height) {
                this.width = width;
                this.height = height;
            }
        }
    }


    public static class Builder {
        private int width;
        private int height;
        private int[][][] strokes;

        public Builder() {        }

        public Builder width(int width) {
            this.width = width;
            return this;
        }
        public Builder height(int height) {
            this.height = height;
            return this;
        }

        Builder setStrokes(int[][][] strokes) {
            this.strokes = strokes;
            return this;
        }

        public EmojiDetectionRequest build() {
            EmojiDetectionRequest edr = new EmojiDetectionRequest();
            edr.options = OPTIONS;
            Request request = new Request();
            request.writingGuide = new Request.WritingGuide(String.valueOf(width),String.valueOf(height));
            request.language = LANGUAGE;
            request.ink = strokes;

            edr.requests = Collections.singletonList(request);

            return edr;
        }

    }

}

File: app/src/main/java/app/we/go/emojidraw/api/EmojiDetectionRequest.kt
package app.we.go.emojidraw.api

import com.google.gson.annotations.SerializedName


class EmojiDetectionRequest(val options: String, val requests: List<Request>) {

    class Request(
        @field:SerializedName("writing_guide")
        val writingGuide: WritingGuide,

        val ink: Array<Array<IntArray>>,

        val language: String? = null) {

        class WritingGuide(@field:SerializedName("writing_area_width")
                           val width: String, @field:SerializedName("writing_area_height")
                           val height: String)

    }

    companion object {

        const val OPTIONS = "enable_pre_space"
        const val LANGUAGE = "emoji_rotated"
    }

    class Builder {
        private var width: Int = 0
        private var height: Int = 0
        private lateinit var strokes: Array<Array<IntArray>>

        fun width(width: Int): Builder {
            this.width = width
            return this
        }

        fun height(height: Int): Builder {
            this.height = height
            return this
        }

        fun setStrokes(strokes: Array<Array<IntArray>>): Builder {
            this.strokes = strokes
            return this
        }

        fun build(): EmojiDetectionRequest {
            requireNotNull(strokes)
            val request = EmojiDetectionRequest.Request(
                writingGuide = EmojiDetectionRequest.Request.WritingGuide(width.toString(), height.toString()),
                language = LANGUAGE,
                ink = strokes)

            return EmojiDetectionRequest(options = OPTIONS, requests = listOf(request))

        }

    }

}





